src/mcts/search.cc

```diff
@@ -451,35 +456,44 @@ EdgeAndNode Search::GetBestChildWithTemperature(Node* parent,
   }
 
   assert(parent->GetChildrenVisits() > 0);
-  std::vector<float> cumulative_sums;
-  float sum = 0.0;
-  const float n_parent = parent->GetN();
+  std::vector<EdgeAndNode> accepted_edges, filtered_edges;
 
   for (auto edge : parent->Edges()) {
     if (parent == root_node_ && !root_limit.empty() &&
         std::find(root_limit.begin(), root_limit.end(), edge.GetMove()) ==
             root_limit.end()) {
       continue;
     }
+
+    if (edge.GetN() > static_cast<unsigned int>(kTempVisitsFloor)) {
+      accepted_edges.push_back(edge);
+    } else {
+      filtered_edges.push_back(edge);
+    }
+  }
+
+  if (accepted_edges.empty()) {
+    accepted_edges = filtered_edges;
+  }
+
+  assert(accepted_edges.size() != 0);
+
+  const float n_parent = parent->GetN();
+  float sum = 0.0;
+  std::vector<float> cumulative_sums;
+  cumulative_sums.reserve(accepted_edges.size());
+
+  for (const auto& edge : accepted_edges) {
     sum += std::pow(edge.GetN() / n_parent, 1 / temperature);
     cumulative_sums.push_back(sum);
   }
 
-  float toss = Random::Get().GetFloat(cumulative_sums.back());
-  int idx =
+  const float toss = Random::Get().GetFloat(cumulative_sums.back());
+  const int idx =
       std::lower_bound(cumulative_sums.begin(), cumulative_sums.end(), toss) -
       cumulative_sums.begin();
 
-  for (auto edge : parent->Edges()) {
-    if (parent == root_node_ && !root_limit.empty() &&
-        std::find(root_limit.begin(), root_limit.end(), edge.GetMove()) ==
-            root_limit.end()) {
-      continue;
-    }
-    if (idx-- == 0) return edge;
-  }
-  assert(false);
-  return {};
+  return accepted_edges.at(idx);
```

Comment:
assert cumulative_sums and accepted_edges has the same size? (Its trivially true, but its an assumption this line is making.)

Reply:
std::vector::at does bounds checking, so it should be the same thing as the assertion. Or am I misunderstanding?

Reply:
Right you are, I always forget since I never use at :P